lib/honeybadger/context_manager.rb

```diff
@@ -21,15 +21,36 @@ def clear!
     # Internal helpers
 
 
-    def set_context(hash)
+    def set_context(hash, &block)
+      local = block_given?
       @mutex.synchronize do
-        @context ||= {}
-        @context.update(Context(hash))
+        @global_context ||= {}
+        @local_context ||= []
+
+        new_context = Context(hash)
+
+        if local
+          @local_context << new_context
+        else
+          @global_context.update(new_context)
+        end
+      end
+
+      if local
+        begin
+          yield
+        ensure
+          @mutex.synchronize { @local_context.pop }
+        end
       end
     end
 
     def get_context
-      @mutex.synchronize { @context }
+      @mutex.synchronize do
+        return nil unless @global_context
+
+        @global_context.merge(@local_context.inject({}, :merge))
```

Comment:
Is there ever a case where `@global_context` will not be nil but `@local_context` will be? I'm concerned about the call to `inject` raising a NoMethodError. Would a guard be useful?  

```suggestion
        @global_context.merge(@local_context.inject({}, :merge)) unless @local_context.nil?
```

Reply:
@stympy I checked and there's no way, they're both always initialized and nullified in the same places, hence why no extra guard. I can still add it if you think it's beneficial